### DISCUSSION TITLE
Displaying states in a dropdown

### DISCUSSION THREAD

**Discussion by ju5t** [spatie/laravel-model-states#189] [Q&A] [answered]
We're using states all over in our application. Thanks for making this!

I have a question about displaying states in a dropdown (e.g. for filtering, or manually changing the state of a model). We're serialising states, so each state has a static `$name` property. We also give each state a `label()` which we display in our views.

How would I go about creating a dropdown where we can use `$name` as the key and the `label()` as it's value?

I know about `getStatesFor` but it's just handing back keys. I think `getStateConfig` goes a long way, but this is a private method. I'm assuming this is done for a good reason.

Any help would be greatly appreciated!

**Comment by brendt** [comment, marked as answer]
There's no built-in support for generating that kind of data, since it's out of scope for this package.

The way we solved it is by adding a `stateOptions(string $stateClass): array` method on a base view model class in our projects. If you're not using view models you could write such a function somewhere else so that it's reusable.

It looks something like this:

```php
    /**
     * @param string|\Spatie\ModelStates\State|array|\Illuminate\Support\Collection $stateOptions
     *
     */
    protected function stateOptions($stateOptions, Model $model): array
    {
        if (is_subclass_of($stateOptions, State::class)) {
            $stateOptions = $stateOptions::all();
        }

        if (is_array($stateOptions)) {
            $stateOptions = collect($stateOptions);
        }

        return $this->options($stateOptions->mapWithKeys(function (string $className) use ($model) {
            /**
             * @var \Spatie\ModelStates\State
             * @var \Spatie\ModelStates\State $state
             */
            $state = new $className($model);

            $label = method_exists($state, 'label')
                ? $state->label()
                : $state::getMorphClass();

            return [$className::getMorphClass() => $label];
        }));
    }
```

If you don't have a concrete model at hand, we'd simply pass `new MyModel()` on the fly.

**Reply by ju5t** [reply]
This solution is perfect! Should've done some more source diving to figure this out myself. Thanks for the help! I guess the solution is now documented here, but perhaps it's worthwhile to add it to the docs? Although that would probably make it in scope again :)

**Reply by brendt** [reply]
Yea I don't want to push people in any direction on how they use it for these kinds of things :)

**Reply by kerrinhardy** [reply]
What should be passed in `stateOptions` method?  What is i the $stateOptions variable?

**Reply by brendt** [reply]
A list of state class names or the abstract state class.
